lib/encryptionmanager.cpp

```diff
@@ -0,0 +1,189 @@
+        }
+        QHash<QString, int> oneTimeKeysToUploadCounts;
+        QHash<QString, int> targetOneTimeKeyCounts
+        {
+            {signedCurve25519Name, qRound(signedKeysProportion * targetKeysNumber)},
+            {curve25519Name, qRound((1-signedKeysProportion) * targetKeysNumber)}
+        };
+};
+
+EncryptionManager::EncryptionManager(const QByteArray &encryptionAccountPickle, float signedKeysProportion, float oneTimeKeyThreshold,
+                                     QObject* parent)
+    : QObject(parent),
+    d(std::make_unique<Private>(std::move(encryptionAccountPickle), std::move(signedKeysProportion), std::move(oneTimeKeyThreshold)))
+{
+
+}
+
+EncryptionManager::~EncryptionManager()
+{
+
+}
+
+void EncryptionManager::uploadIdentityKeys(Connection* connection)
+{
+    DeviceKeys deviceKeys;
+
+    deviceKeys.userId = connection->userId();
+    deviceKeys.deviceId = connection->deviceId();
+
+    QStringList algorithms = { olmCurve25519AesSha256AlgoName,
+                               megolmV1AesShaAlgoName};
+    deviceKeys.algorithms = algorithms;
+
+    QHash<QString, QString> keys;
+    keys.insert(curve25519Name + QStringLiteral(":") + connection->deviceId(),
+                d->olmAccount->curve25519IdentityKey());
+    keys.insert(ed25519Name + QStringLiteral(":") + connection->deviceId(),
+                d->olmAccount->ed25519IdentityKey());
+    deviceKeys.keys = keys;
+
+    QJsonObject message;
+    JsonObjectConverter<DeviceKeys>::dumpTo(message, deviceKeys);
+
+    QHash<QString, QHash<QString, QString>> signatures;
+    QHash<QString, QString> signatureDict;
+    signatureDict.insert(ed25519Name + QStringLiteral(":") + connection->deviceId(),
+                     d->olmAccount->sign(message));
+    signatures.insert(connection->userId(),signatureDict);
+
+    connect(d->uploadIdentityKeysJob, &BaseJob::success, this, [this] {
+        d->setOneTimeKeyCounts(d->uploadIdentityKeysJob->oneTimeKeyCounts());
+        qDebug() << QString("Uploaded identity keys.");
+    });
+    d->uploadIdentityKeysJob = connection->callApi<UploadKeysJob>(deviceKeys);
+}
+
+void EncryptionManager::uploadOneTimeKeys(Connection* connection, bool forceUpdate)
+{
+    if (forceUpdate || d->oneTimeKeyCounts.isEmpty())
+    {
+        auto job = connection->callApi<UploadKeysJob>();
+        connect(job, &BaseJob::success, this, [job,this] {
+            d->setOneTimeKeyCounts(job->oneTimeKeyCounts());
+        });
+
+    }
+
+    int signedKeysToUploadCount = d->oneTimeKeysToUploadCounts.value(signedCurve25519Name, 0);
+    int unsignedKeysToUploadCount = d->oneTimeKeysToUploadCounts.value(curve25519Name, 0);
+
+    d->olmAccount->generateOneTimeKeys(signedKeysToUploadCount + unsignedKeysToUploadCount);
+
+    QHash<QString, QVariant> oneTimeKeys = {};
+    const QVariantMap& olmAccountCurve25519OneTimeKeys = d->olmAccount->curve25519OneTimeKeys();
+    for (int i = 0; i < olmAccountCurve25519OneTimeKeys.count(); ++i)
```

Comment:
That's a pretty 90's way of iterating over a container :) I understand you need a counter but you can count and iterate separately. The reason why it matters is that `keys()` below is _copying_ keys to a temporary - on every iteration. So better use `begin()` and `end()` for iteration, increment a separate counter before completion of the loop, and use `key()`and `value()` instead of `keys().at()` and `value(keyId)`.

...oh, and by the way, think of using `oneTimeKeys.size()` instead of  the separate counter.